semantic-analysis/src/Analysis/Syntax.hs

```diff
@@ -0,0 +1,181 @@
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE UndecidableInstances #-}
+module Analysis.Syntax
+( Syntax(..)
+  -- * Pretty-printing
+, Print(..)
+  -- * Abstract interpretation
+, eval0
+, eval
+, Interpret(..)
+  -- * Macro-expressible syntax
+, let'
+  -- * Parsing
+, parseFile
+, parseGraph
+, parseNode
+) where
+
+import           Analysis.Effect.Domain
+import           Analysis.Effect.Env (Env, bind)
+import           Analysis.Effect.Store
+import           Analysis.Name (Name, formatName, nameI)
+import           Control.Applicative (Alternative(..), liftA3)
+import           Control.Effect.Labelled
+import           Control.Monad (guard)
+import qualified Data.Aeson as A
+import qualified Data.Aeson.Internal as A
+import qualified Data.Aeson.Parser as A
+import qualified Data.Aeson.Types as A
+import qualified Data.ByteString.Lazy as B
+import           Data.Function (fix)
+import qualified Data.IntMap as IntMap
+import           Data.Monoid (First(..))
+import           Data.Text (Text, pack, unpack)
+import qualified Data.Vector as V
+
+class Syntax rep where
+  iff :: rep -> rep -> rep -> rep
+  noop :: rep
+
+  bool :: Bool -> rep
+  string :: Text -> rep
+
+  throw :: rep -> rep
+
+  let_ :: Name -> rep -> (rep -> rep) -> rep
+
+
+-- Pretty-printing
+
+newtype Print = Print { print_ :: ShowS }
+
+instance Show Print where
+  showsPrec _ = print_
+
+instance Semigroup Print where
+  Print a <> Print b = Print (a . b)
+
+instance Monoid Print where
+  mempty = Print id
+
+instance Syntax Print where
+  iff c t e = parens (str "iff" <+> c <+> str "then" <+> t <+> str "else" <+> e)
+  noop = parens (str "noop")
+
+  bool b = parens (str (if b then "true" else "false"))
+  string = parens . text
+
+  throw e = parens (str "throw" <+> e)
+
+  let_ n v b = parens (str "let" <+> name n <+> char '=' <+> v <+> str "in" <+> b (name n))
+
+str :: String -> Print
+str = Print . showString
+
+text :: Text -> Print
+text = str . unpack
```

Comment:
You could unify `text` and `str` if you used `IsString`, it occurs to me

Reply:
True, but I like the explicitness here. Also `Print` is basically debug code, not something I expect will see much use or maintenance 😊